File: src/code_generator.cpp
#include "code_generator.h"
#include <sstream>
#include <iostream>
#include "ast.h"

std::string CodeGenerator::generate(ASTNode* node) {
    std::stringstream ss;
    generateNode(node, ss);
    return ss.str();
}

void CodeGenerator::generateNode(ASTNode* node, std::stringstream& ss) {
    if (auto* varDecl = dynamic_cast<VariableDeclarationNode*>(node)) {
        generateVariableDeclaration(varDecl, ss);
    } else if (auto* funcDecl = dynamic_cast<FunctionDeclarationNode*>(node)) {
        generateFunctionDeclaration(funcDecl, ss);
    } else if (auto* ifStmt = dynamic_cast<IfStatementNode*>(node)) {
        generateIfStatement(ifStmt, ss);
    } else if (auto* forStmt = dynamic_cast<ForStatementNode*>(node)) {
        generateForStatement(forStmt, ss);
    } else if (auto* whileStmt = dynamic_cast<WhileStatementNode*>(node)) {
        generateWhileStatement(whileStmt, ss);
    } else if (auto* returnStmt = dynamic_cast<ReturnStatementNode*>(node)) {
        generateReturnStatement(returnStmt, ss);
    } else if (auto* classDecl = dynamic_cast<ClassDeclarationNode*>(node)) {
        generateClassDeclaration(classDecl, ss);
    } else if (auto* printStmt = dynamic_cast<PrintStatementNode*>(node)) {
        generatePrintStatement(printStmt, ss);
    } else if (auto* exprStmt = dynamic_cast<ExpressionStatementNode*>(node)) {
        generateExpressionStatement(exprStmt, ss);
    } else if (auto* binaryExpr = dynamic_cast<BinaryExpressionNode*>(node)) {
        generateBinaryExpression(binaryExpr, ss);
    } else if (auto* unaryExpr = dynamic_cast<UnaryExpressionNode*>(node)) {
        generateUnaryExpression(unaryExpr, ss);
    } else if (auto* literal = dynamic_cast<LiteralNode*>(node)) {
        generateLiteral(literal, ss);
    } else if (auto* variable = dynamic_cast<VariableNode*>(node)) {
        generateVariable(variable, ss);
    } else if (auto* grouping = dynamic_cast<GroupingNode*>(node)) {
        generateGrouping(grouping, ss);
    } else if (auto* assignment = dynamic_cast<AssignmentNode*>(node)) {
        generateAssignment(assignment, ss);
    }
}

void CodeGenerator::generateVariableDeclaration(VariableDeclarationNode* node, std::stringstream& ss) {
    if (!node->initializer) {
        std::cerr << "Error: Missing initializer in variable declaration" << std::endl;
        return;
    }
    ss << (node->type == TokenType::LET ? "auto " : "const auto ") << node->name.lexeme << " = ";
    generateNode(node->initializer, ss);
    ss << ";\n";
}

void CodeGenerator::generateFunctionDeclaration(FunctionDeclarationNode* node, std::stringstream& ss) {
    ss << "auto " << node->name.lexeme << "(";
    for (size_t i = 0; i < node->parameters.size(); ++i) {
        ss << "auto " << node->parameters[i].lexeme;
        if (i < node->parameters.size() - 1) {
            ss << ", ";
        }
    }
    ss << ") {\n";
    for (auto* stmt : node->body) {
        generateNode(stmt, ss);
    }
    ss << "}\n";
}

void CodeGenerator::generateIfStatement(IfStatementNode* node, std::stringstream& ss) {
    ss << "if (";
    generateNode(node->condition, ss);
    ss << ") {\n";
    generateNode(node->thenBranch, ss);
    ss << "}\n";
    if (node->elseBranch) {
        ss << "else {\n";
        generateNode(node->elseBranch, ss);
        ss << "}\n";
    }
}

void CodeGenerator::generateForStatement(ForStatementNode* node, std::stringstream& ss) {
    ss << "for (";
    if (node->initializer) {
        generateNode(node->initializer, ss);
    }
    ss << "; ";
    if (node->condition) {
        generateNode(node->condition, ss);
    }
    ss << "; ";
    if (node->increment) {
        generateNode(node->increment, ss);
    }
    ss << ") {\n";
    generateNode(node->body, ss);
    ss << "}\n";
}

void CodeGenerator::generateWhileStatement(WhileStatementNode* node, std::stringstream& ss) {
    ss << "while (";
    generateNode(node->condition, ss);
    ss << ") {\n";
    generateNode(node->body, ss);
    ss << "}\n";
}

void CodeGenerator::generateReturnStatement(ReturnStatementNode* node, std::stringstream& ss) {
    ss << "return ";
    if (node->value) {
        generateNode(node->value, ss);
    }
    ss << ";\n";
}

void CodeGenerator::generateClassDeclaration(ClassDeclarationNode* node, std::stringstream& ss) {
    ss << "class " << node->name.lexeme << " {\npublic:\n";
    for (auto* method : node->methods) {
        generateNode(method, ss);
    }
    ss << "};\n";
}

void CodeGenerator::generatePrintStatement(PrintStatementNode* node, std::stringstream& ss) {
    ss << "std::cout << ";
    generateNode(node->value, ss);
    ss << " << std::endl;\n";
}

void CodeGenerator::generateExpressionStatement(ExpressionStatementNode* node, std::stringstream& ss) {
    generateNode(node->expression, ss);
    ss << ";\n";
}

void CodeGenerator::generateBinaryExpression(BinaryExpressionNode* node, std::stringstream& ss) {
    generateNode(node->left, ss);
    ss << " " << node->op.lexeme << " ";
    generateNode(node->right, ss);
}

void CodeGenerator::generateUnaryExpression(UnaryExpressionNode* node, std::stringstream& ss) {
    ss << node->op.lexeme;
    generateNode(node->right, ss);
}

void CodeGenerator::generateLiteral(LiteralNode* node, std::stringstream& ss) {
    ss << node->value;
}

void CodeGenerator::generateVariable(VariableNode* node, std::stringstream& ss) {
    ss << node->name.lexeme;
}

void CodeGenerator::generateGrouping(GroupingNode* node, std::stringstream& ss) {
    ss << "(";
    generateNode(node->expression, ss);
    ss << ")";
}

void CodeGenerator::generateAssignment(AssignmentNode* node, std::stringstream& ss) {
    ss << node->name.lexeme << " = ";
    generateNode(node->value, ss);
}

File: src/code_generator.ts
import { ASTNode, VariableDeclarationNode, FunctionDeclarationNode, IfStatementNode, ForStatementNode, WhileStatementNode, ReturnStatementNode, ClassDeclarationNode, PrintStatementNode, ExpressionStatementNode, BinaryExpressionNode, UnaryExpressionNode, LiteralNode, VariableNode, GroupingNode, AssignmentNode } from './ast';

export class CodeGenerator {
    generate(node: ASTNode): string {
        let code = '';
        this.generateNode(node, code);
        return code;
    }

    private generateNode(node: ASTNode, code: string): void {
        if (node instanceof VariableDeclarationNode) {
            this.generateVariableDeclaration(node, code);
        } else if (node instanceof FunctionDeclarationNode) {
            this.generateFunctionDeclaration(node, code);
        } else if (node instanceof IfStatementNode) {
            this.generateIfStatement(node, code);
        } else if (node instanceof ForStatementNode) {
            this.generateForStatement(node, code);
        } else if (node instanceof WhileStatementNode) {
            this.generateWhileStatement(node, code);
        } else if (node instanceof ReturnStatementNode) {
            this.generateReturnStatement(node, code);
        } else if (node instanceof ClassDeclarationNode) {
            this.generateClassDeclaration(node, code);
        } else if (node instanceof PrintStatementNode) {
            this.generatePrintStatement(node, code);
        } else if (node instanceof ExpressionStatementNode) {
            this.generateExpressionStatement(node, code);
        } else if (node instanceof BinaryExpressionNode) {
            this.generateBinaryExpression(node, code);
        } else if (node instanceof UnaryExpressionNode) {
            this.generateUnaryExpression(node, code);
        } else if (node instanceof LiteralNode) {
            this.generateLiteral(node, code);
        } else if (node instanceof VariableNode) {
            this.generateVariable(node, code);
        } else if (node instanceof GroupingNode) {
            this.generateGrouping(node, code);
        } else if (node instanceof AssignmentNode) {
            this.generateAssignment(node, code);
        }
    }

    private generateVariableDeclaration(node: VariableDeclarationNode, code: string): void {
        if (!node.initializer) {
            throw new Error('Missing initializer in variable declaration');
        }
        code += `${node.type === 'LET' ? 'let' : 'const'} ${node.name.lexeme} = `;
        this.generateNode(node.initializer, code);
        code += ';\n';
    }

    private generateFunctionDeclaration(node: FunctionDeclarationNode, code: string): void {
        code += `function ${node.name.lexeme}(`;
        for (let i = 0; i < node.parameters.length; i++) {
            code += node.parameters[i].lexeme;
            if (i < node.parameters.length - 1) {
                code += ', ';
            }
        }
        code += ') {\n';
        for (const stmt of node.body) {
            this.generateNode(stmt, code);
        }
        code += '}\n';
    }

    private generateIfStatement(node: IfStatementNode, code: string): void {
        code += 'if (';
        this.generateNode(node.condition, code);
        code += ') {\n';
        this.generateNode(node.thenBranch, code);
        code += '}\n';
        if (node.elseBranch) {
            code += 'else {\n';
            this.generateNode(node.elseBranch, code);
            code += '}\n';
        }
    }

    private generateForStatement(node: ForStatementNode, code: string): void {
        code += 'for (';
        if (node.initializer) {
            this.generateNode(node.initializer, code);
        }
        code += '; ';
        if (node.condition) {
            this.generateNode(node.condition, code);
        }
        code += '; ';
        if (node.increment) {
            this.generateNode(node.increment, code);
        }
        code += ') {\n';
        this.generateNode(node.body, code);
        code += '}\n';
    }

    private generateWhileStatement(node: WhileStatementNode, code: string): void {
        code += 'while (';
        this.generateNode(node.condition, code);
        code += ') {\n';
        this.generateNode(node.body, code);
        code += '}\n';
    }

    private generateReturnStatement(node: ReturnStatementNode, code: string): void {
        code += 'return ';
        if (node.value) {
            this.generateNode(node.value, code);
        }
        code += ';\n';
    }

    private generateClassDeclaration(node: ClassDeclarationNode, code: string): void {
        code += `class ${node.name.lexeme} {\n`;
        for (const method of node.methods) {
            this.generateNode(method, code);
        }
        code += '}\n';
    }

    private generatePrintStatement(node: PrintStatementNode, code: string): void {
        code += 'console.log(';
        this.generateNode(node.value, code);
        code += ');\n';
    }

    private generateExpressionStatement(node: ExpressionStatementNode, code: string): void {
        this.generateNode(node.expression, code);
        code += ';\n';
    }

    private generateBinaryExpression(node: BinaryExpressionNode, code: string): void {
        this.generateNode(node.left, code);
        code += ` ${node.op.lexeme} `;
        this.generateNode(node.right, code);
    }

    private generateUnaryExpression(node: UnaryExpressionNode, code: string): void {
        code += node.op.lexeme;
        this.generateNode(node.right, code);
    }

    private generateLiteral(node: LiteralNode, code: string): void {
        code += node.value;
    }

    private generateVariable(node: VariableNode, code: string): void {
        code += node.name.lexeme;
    }

    private generateGrouping(node: GroupingNode, code: string): void {
        code += '(';
        this.generateNode(node.expression, code);
        code += ')';
    }

    private generateAssignment(node: AssignmentNode, code: string): void {
        code += `${node.name.lexeme} = `;
        this.generateNode(node.value, code);
    }
}
